controllers/key_controller.go

```diff
@@ -0,0 +1,274 @@
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-github/v28/github"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1alpha1 "go.hein.dev/github-controller/api/v1alpha1"
+	"go.hein.dev/github-controller/git"
+	"go.hein.dev/github-controller/keygen"
+)
+
+var (
+	keyFinalizerName = "key.finalizers.github.go.hein.dev"
+)
+
+// KeyReconciler reconciles a Key object
+type KeyReconciler struct {
+	client.Client
+	Log          logr.Logger
+	Scheme       *runtime.Scheme
+	GitClient    git.Client
+	ActualDelete bool
+}
+
+// +kubebuilder:rbac:groups=github.go.hein.dev,resources=keys,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=github.go.hein.dev,resources=keys/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=,resources=secrets,verbs=get;list;watch;create
+
+func (r *KeyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var requeueAfter = 2 * time.Second
+	ctx := context.Background()
+	log := r.Log.WithValues("key", req.NamespacedName)
+
+	var key v1alpha1.Key
+	if err := r.Client.Get(ctx, req.NamespacedName, &key); err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// handle finalizers before any other reconcile logic can fail
+	if !key.ObjectMeta.DeletionTimestamp.IsZero() &&
+		containsString(key.GetFinalizers(), keyFinalizerName) {
+		log.Info("handle deletion", "name", key.Name)
+		if err := r.handleDeletion(ctx, &key); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// fetch or create the desired Secret
+	// the child secret defaults to the same namespace/name as the key
+	secretRef := req.NamespacedName
+	// ensure the desired Secret is referenced by the desired namespace/name
+	if key.Spec.SecretTemplate.NameOverride != "" {
+		secretRef.Name = key.Spec.SecretTemplate.NameOverride
+	}
+	if key.Spec.SecretTemplate.TargetNamespace != "" {
+		secretRef.Namespace = key.Spec.SecretTemplate.TargetNamespace
+	}
+	log = log.WithValues("secret", secretRef)
+	var secret corev1.Secret
+	if err := r.Client.Get(ctx, secretRef, &secret); err != nil {
+		if errors.IsNotFound(err) {
+			// any previous secret no longer exists -- optimistically delete the old key
+			if key.Status.GitHubKeyID != 0 {
+				r.updateKeyStatus(ctx, &key, v1alpha1.DeletingStatus)
+				err = r.GitClient.DeleteKey(ctx, key.Status.GitHubOrganization, key.Status.GitHubRepository, key.Status.GitHubKeyID)
+				if err != nil {
+					return ctrl.Result{RequeueAfter: requeueAfter}, err
+				}
+			}
+
+			r.updateKeyStatus(ctx, &key, v1alpha1.CreatingStatus)
+			log.Info("referenced secret does not exist, generating new key")
```

Comment:
I wonder if k8s events would be better instead of logs so the user would know what is happening. wdyt?